### ISSUE TITLE
[OpenMP] Unsound deduplication: omp_get_thread_limit

### ISSUE THREAD

**Issue by MattPD** [llvm/llvm-project#74673]
Here's a simple example:
```
#include <omp.h>
#include <stdio.h>

int main()
{
	#pragma omp target thread_limit(4)
	{
		printf("\n1:target thread_limit: %d\n", omp_get_thread_limit());
	}

	#pragma omp target thread_limit(3)
	{
		printf("\n2:target thread_limit: %d\n", omp_get_thread_limit());
	}
	return 0;
}
```

GCC-compiled binary execution: https://gcc.godbolt.org/z/Pjv3TWoTq
```
1:target thread_limit: 4
2:target thread_limit: 3
```

Clang/LLVM-compiled binary execution: https://clang.godbolt.org/z/zdPbrdMPn
```
1:target thread_limit: 4
2:target thread_limit: 4
```

By my reading of the OpenMP spec GCC does the right thing here; cf. <https://www.openmp.org/spec-html/5.2/openmpse12.html#x34-330002.4>:
> If a target construct with a thread_limit clause is encountered, the thread-limit-var ICV from the data environment of the generated initial task is instead set to an implementation deﬁned value between one and the value speciﬁed in the clause. 

The common subexpression elimination (CSE) of the second call to `omp_get_thread_limit` by LLVM does not seem to be correct, as it's not an available expression at any program point(s) (in the scope of the clause in question) after the second target construct with a `thread_limit` clause is encountered.

Compiling with `-Rpass=openmp-opt -Rpass-analysis=openmp-opt -Rpass-missed=openmp-opt` we have:
https://clang.godbolt.org/z/G7dfhP7jh
```
<source>:8:42: remark: OpenMP runtime call omp_get_thread_limit deduplicated. [OMP170] [-Rpass=openmp-opt]
    8 |                 printf("\n1:target thread_limit: %d\n",omp_get_thread_limit());
      |                                                        ^
```

OMP170 has the following explanation: https://openmp.llvm.org/remarks/OMP170.html

> This optimization remark indicates that a call to an OpenMP runtime call was replaced with the result of an existing one. This occurs when the compiler knows that the result of a runtime call is immutable. Removing duplicate calls is done by replacing all calls to that function with the result of the first call. This cannot be done automatically by the compiler because the implementations of the OpenMP runtime calls live in a separate library the compiler cannot see.
This optimization will trigger for known OpenMP runtime calls whose return value will not change.

What's interesting is that the above example for OMP170 shows `omp_get_thread_limit`, too (notably without any intervening changes to thread_limit, like in the reproducer in question).

At the same time I do not believe we have an analysis checking whether this precondition holds here: "This occurs when the compiler knows that the result of a runtime call is immutable."

I presume that in order to determine the "immutable" part we'd need a form of flow-sensitive analysis that would enable us to check for and implement the invalidation ("kill" in the dataflow analysis sense) of the-result-of-the-call-to`omp_get_thread_limit` as an available expression (or other functions listed in `DeduplicableRuntimeCallIDs`)--in other words, anything that would allow us to express that prior calls (like `__kmpc_set_thread_limit`) affect the result of subsequent calls of `omp_get_thread_limit`; cf. the commit implementing the ICV, https://github.com/llvm/llvm-project/commit/d7b12004bd7d6d9a592f1773101cbedd9daf8492

AFAICT, such analysis doesn't appear to exist in the original patch introducing deduplication, either:

- https://github.com/llvm/llvm-project/commit/9548b74a831ea005649465797f359e0521f3b8a9
- https://reviews.llvm.org/D69930

> This initial patch deduplicates `__kmpc_global_thread_num` and `omp_get_thread_num` calls. We can also identify arguments that are equivalent to such a call result and use it instead. Later we can determine "gtid" arguments based on the use in kernel functions etc.

In particular, cf. `deduplicateRuntimeCalls(Function &F, RuntimeFunctionInfo &RFI, Value *ReplVal = nullptr)` in "llvm/lib/Transforms/IPO/OpenMPOpt.cpp". AFAICT, the checked precondition boils down to the presence of _use_:
```
    auto &Uses = RFI.UsesMap[&F];
    if (Uses.size() + (ReplVal != nullptr) < 2)
      return false;
```

If the required flow-sensitive analysis does exists, we could perhaps invalidate the relevant entry (if that's supported?) in `OMPInformationCache::RuntimeFunctionInfo &RFI`; cf. the use of `OMPInfoCache.RFIs` in `deduplicateRuntimeCalls`
https://github.com/llvm/llvm-project/blob/2b76e20ea782790a78ec58d5f94ce88a173bab7f/llvm/lib/Transforms/IPO/OpenMPOpt.cpp#L1467

FWIW, there is an optimization (constant folding) that seems somewhat more precise (i.e., checking for the presence of the `OMPThreadLimitClause` such that `ThreadLimit` from that clause `isIntegerConstantExpr`), cf.

- [OpenMP] Creating the `omp_target_num_teams` and `omp_target_thread_limit` attributes to outlined functions
- https://reviews.llvm.org/D106298
- https://github.com/llvm/llvm-project/commit/0276db14167b9348904322084e7fc1a04cc72452

The relevant logic being this bit:
```
const Expr *CGOpenMPRuntime::getNumThreadsExprForTargetDirective(
    CodeGenFunction &CGF, const OMPExecutableDirective &D,
    int32_t &DefaultVal)
    . . .
    Expr *ThreadLimit = nullptr;
    Expr *NumThreads = nullptr;
    if (D.hasClausesOfKind<OMPThreadLimitClause>()) {
      const auto *ThreadLimitClause = D.getSingleClause<OMPThreadLimitClause>();
      ThreadLimit = ThreadLimitClause->getThreadLimit();
      if (ThreadLimit->isIntegerConstantExpr(CGF.getContext()))
        if (auto Constant =
                ThreadLimit->getIntegerConstantExpr(CGF.getContext()))
          DefaultVal = Constant->getExtValue();
    }
```

This is further refined in [OpenMP] Folding threadLimit and numThreads when single value in kernels, https://reviews.llvm.org/D106033

That being said, this is a more special case of _constant_ folding, so there may not be a comparable analysis for (non-constant) available expressions in general.

I do not believe we'd lose the constant folding optimization if we were to revert deduplication, and if the deduplication has only been implemented for immutable functions I don't see how it can be correct (with the particular case of ThreadLimit being already covered by the aforementioned constant folding, the actually constant case appears to be subsumed, too).

One idea for the fix in this context: Treat `omp_get_thread_limit` exactly like `omp_get_max_threads`.

What this means in practice is to remove it from `DeduplicableRuntimeCallIDs`, effectively reverting the addition in this commit (noting that `omp_get_max_threads` is not present in `DeduplicableRuntimeCallIDs`, so it's possible this addition was incorrect in the first place):

- [OpenMP][Opt] Annotate known runtime functions and deduplicate more,
- https://github.com/llvm/llvm-project/commit/e28936f6137c5a9c4f7673e248c192a9811543b6#diff-de101c82aff66b2bda2d1f53fde3dde7b0d370f14f1ff37b7919ce38531230dfR123

Another idea--if this is possible and correct here--would be to special-case the particular case of `omp_get_thread_limit` (possibly applicable to other similar functions?) in `deduplicateRuntimeCalls` to only perform the deduplication within a more limited scope (perhaps a basic block--or basic blocks--corresponding to the nested scope of the applicable target construct with a thread_limit clause that's in effect?).

Thoughts?

**Comment by MattPD** [comment]
Hey, @jdoerfert, I'm wondering, would you happen to have thoughts on this?

If I understand correctly, the deduplication used to be legal when originally added in <https://github.com/llvm/llvm-project/commit/e28936f6137c5a9c4f7673e248c192a9811543b6#diff-de101c82aff66b2bda2d1f53fde3dde7b0d370f14f1ff37b7919ce38531230dfR123>, as the result was immutable.

However, now that we have thread_limit clause, we no longer have immutability.

Does removing `omp_get_thread_limit` from the `DeduplicableRuntimeCallIDs` array sound like the way to go?
If so I can push a quick patch.

I understand it's possible that we may lose potential optimization opportunities, however, at the same time I believe we generally put correctness first (so it may still make sense to disable while we're working on the fix, if there's any?).

FWIW, LLVM IR before/after: https://clang.godbolt.org/z/GzbsP86WK

Interestingly, looks like it doesn't apply when it's outlined as a separate function, say, when using parallel construct: https://clang.godbolt.org/z/x6db9c39z

If it's possible to limit the scope of the deduplication (so it doesn't cross the `omp target` scope boundaries), or outline target scope similarly to parallel (not clear on the pros/cons of this approach), or implement flow-sensitive tracking invalidating the results (not clear how if this is a viable option), we can perhaps restore this in the future.

**Comment by MattPD** [comment]
Opened a PR with a proposed fix: #79524

**Comment by MattPD** [comment]
Closing: PR merged in commit [88e31f6](https://github.com/llvm/llvm-project/commit/88e31f64a034ec6dead2106016ee5b797674edb0).
